docs/plugins/npe2_manifest_specification.md

```diff
@@ -0,0 +1,578 @@
+
+### Optional
+
+- **engine**: A SemVer compatible version string matching the versions of the
+  plugin engine that the extension is compatible with. Example: "0.1.0".
+- **display_name**: User-facing text to display as the name of this plugin.
+  Example: `napari SVG`. Must be 3-40 characters long, containing
+  printable word characters, and must not begin or end with an underscore,
+  white space, or non-word character.
+- **entry_point**: The module containing the `activate()` function. Example:
+  `foo.bar.baz`. This should be a fully qualified module string.
+
+(commands)=
+
+## Commands
+
+**command** is a python function associated with an id. The **id** is
+a unique identifier to which other contributions, like readers, can refer.
+
+### Required fields
+
+- **id** An identifier used to reference this command within this plugin.
+- **title** A description of the command. This might be used, for example,
+  when searching in a command palette. Examples: "Generate lily sample",
+  "Read tiff image", "Open gaussian blur widget".
+
+### Optional fields
+
+- **python_name** Fully qualified name to callable python object implementing
+  this command. This usually takes the form of
+  `{obj.__module__}:{obj.__qualname__}` (e.g.
+  `my_package.a_module:some_function`). For a command to be useful, this
+  field must be specified,. If it isn't specified here it should be done
+  dynamically inside the plugin's 'activate()' function.
+
+### Example
+
+```yaml
+name: my-plugin
+contribution:
+  commands:
+    - id: my-plugin.publish_paper
+      title: Do experiments, analysis, write paper, and submit
+      python_name: my_plugin:publish_func
+```
+
+### Dynamic command registration
+
+Commands can be registered dynamically by the plugin's `activate()` function.
+For example,
+
+```python
+# inside my_plugin/__init__.py
+from npe2 import PluginContext
+def activate(context: PluginContext):
+    @context.register_command("my_plugin.hello_world")
+    def _hello():
+        ...
+
+    context.register_command("my_plugin.another_command", lambda: print("yo!"))
+```
+
+The `activate()` function is called when the plugin is loaded. It is found in
+the `entry_point` specified in the plugin manifest.
+
+## Readers
+
+### Required fields
+
+- **command** Identifier of the _command_ to execute.
+- **filename_patterns** List of filename patterns (for fnmatch) that this
+  reader can accept. Reader will be tried only if `fnmatch(filename, pattern) == True`.
+
+### Optional fields
+
+- **accepts_directories** If true, the reader will accept paths to directories
+  for reading data.
+
+### Example
+
+```yaml
+name: my-reader
+contribution:
+  commands:
+    - id: my-reader.read_npy_image
+      title: Open an npy file as an image
+      python_name: my_reader.load_as_image
+  readers:
+    - command: my-reader.read_npy_image
+      accepts_directories: false
+      filename_patterns: ["*.npy"]
+```
+
+### Calling convention
```

Comment:
I don't remember seeing the phrase "calling convention" before this point in the document - maybe it's worth adding a sentence near the top about what a calling convention is?

Reply:
this is a great point and gets to some of your other questions about how commands work.  I'll add some text around this in a follow up pr.